File: Museo/Usuario.py
from abc import ABC, abstractmethod


class Usuario(ABC):
    """Clase abstracta que representa los usuarios del sistema, a partir de esta clase creamos los usuarios."""

    def __init__(
        self,
        nombre_usuario: str,
        contrasena: str
        ) -> None:
        
        self._nombre_usuario = nombre_usuario
        self._contrasena = contrasena
        self._autenticado = False

    @property
    def nombre_usuario(self) -> str:
        return self._nombre_usuario

    @property
    def autenticado(self) -> bool:
        return self._autenticado

    def autenticar(self, contrasena: str) -> bool:
        """Verifica la contraseña y marca al usuario como autenticado."""
        self._autenticado = self._contrasena == contrasena
        return self._autenticado

    def cerrar_sesion(self) -> None:
        self._autenticado = False

    @abstractmethod
    def get_rol(self) -> str:
        """Retorna el rol del usuario en el sistema."""

File: 2.2Museo_Java/Clases_Museo/Usuario.java
package Clases_Museo;

/**
 * Clase abstracta base para todos los usuarios del sistema Museo.
 *
 * Almacena las credenciales y provee autenticacion y cierre de sesion
 * que heredan Visitante, AgenteCatalogo, AgenteRestaurador y DirectorMuseo.
 */
public abstract class Usuario {

    // Atributos protegidos para acceso desde las subclases
    protected String nombreUsuario;
    protected String contrasena;
    protected boolean autenticado;

    // --- Constructor ---

    /**
     * Inicializa el usuario con credenciales y sesion cerrada.
     *
     * @param nombreUsuario Nombre de usuario del sistema.
     * @param contrasena    Contrasena del usuario.
     */
    public Usuario(String nombreUsuario, String contrasena) {
        this.nombreUsuario = nombreUsuario;
        this.contrasena    = contrasena;
        this.autenticado   = false;
    }

    // --- Getters ---

    /** @return El nombre de usuario. */
    public String getNombreUsuario() { return nombreUsuario; }

    /** @return true si el usuario tiene sesion activa. */
    public boolean isAutenticado() { return autenticado; }

    // --- Setter con validacion ---

    /**
     * Actualiza la contrasena del usuario validando que no este vacia.
     *
     * @param nueva Nueva contrasena a establecer.
     */
    public void setContrasena(String nueva) {
        if (nueva == null || nueva.isBlank()) {
            throw new IllegalArgumentException("La contrasena no puede estar vacia.");
        }
        this.contrasena = nueva;
    }

    // --- Metodos ---

    /**
     * Verifica la contrasena y marca al usuario como autenticado.
     *
     * @param contrasena Contrasena a verificar.
     * @return true si la contrasena es correcta.
     */
    public boolean autenticar(String contrasena) {
        this.autenticado = this.contrasena.equals(contrasena);
        return this.autenticado;
    }

    /** Cierra la sesion del usuario. */
    public void cerrarSesion() {
        this.autenticado = false;
    }

    // --- Metodo abstracto ---

    /** @return El rol del usuario en el sistema. */
    public abstract String getRol();

    @Override
    public String toString() {
        return "Usuario(" + nombreUsuario + ", rol=" + getRol() + ")";
    }
}
